provider/kubernetes/kubernetes.go

```diff
@@ -0,0 +1,429 @@
+}
+
+func checkStringQuoteValidity(value string) error {
+	_, err := strconv.Unquote(`"` + value + `"`)
+	return err
+}
+
+func loadService(client Client, namespace string, backend v1beta1.IngressBackend) (*config.Service, error) {
+	service, exists, err := client.GetService(namespace, backend.ServiceName)
+	if err != nil {
+		return nil, err
+	}
+
+	if !exists {
+		return nil, errors.New("service not found")
+	}
+
+	var servers []config.Server
+	var portName string
+	var portSpec corev1.ServicePort
+	var match bool
+	for _, p := range service.Spec.Ports {
+		if (backend.ServicePort.Type == intstr.Int && backend.ServicePort.IntVal == p.Port) ||
+			(backend.ServicePort.Type == intstr.String && backend.ServicePort.StrVal == p.Name) {
+			portName = p.Name
+			portSpec = p
+			match = true
+			break
+		}
+	}
+
+	if !match {
+		return nil, errors.New("service port not found")
+	}
+
+	if service.Spec.Type == corev1.ServiceTypeExternalName {
+		servers = append(servers, config.Server{
+			URL:    fmt.Sprintf("http://%s:%d", service.Spec.ExternalName, portSpec.Port),
+			Weight: 1,
+		})
+	} else {
+		endpoints, endpointsExists, endpointsErr := client.GetEndpoints(namespace, backend.ServiceName)
+		if endpointsErr != nil {
+			return nil, endpointsErr
+		}
+
+		if !endpointsExists {
+			return nil, errors.New("endpoints not found")
+		}
+
+		if len(endpoints.Subsets) == 0 {
+			return nil, errors.New("subset not found")
+		}
+
+		var port int32
+		for _, subset := range endpoints.Subsets {
+
+			for _, p := range subset.Ports {
+				if portName == p.Name {
+					port = p.Port
+					break
+				}
+			}
+
+			if port == 0 {
+				return nil, errors.New("cannot define a port")
+			}
+
+			protocol := "http"
+			if port == 443 || portName == "https" {
+				protocol = "https"
+			}
+
+			for _, addr := range subset.Addresses {
+				servers = append(servers, config.Server{
+					URL:    fmt.Sprintf("%s://%s:%d", protocol, addr.IP, port),
+					Weight: 1,
+				})
+			}
+		}
+	}
+
+	return &config.Service{
+		LoadBalancer: &config.LoadBalancerService{
+			Servers:        servers,
+			Method:         "wrr",
+			PassHostHeader: true,
+		},
+	}, nil
+}
+
+func (p *Provider) loadConfigurationFromIngresses(ctx context.Context, client Client) *config.Configuration {
+	conf := &config.Configuration{
+		Routers:     map[string]*config.Router{},
+		Middlewares: map[string]*config.Middleware{},
+		Services:    map[string]*config.Service{},
+	}
+
+	ingresses := client.GetIngresses()
+
+	tlsConfigs := make(map[string]*tls.Configuration)
+	for _, ingress := range ingresses {
+		ctx = log.With(ctx, log.Str("ingress", ingress.Name), log.Str("namespace", ingress.Namespace))
+
+		if !shouldProcessIngress(p.IngressClass, ingress.Annotations[annotationKubernetesIngressClass]) {
+			continue
+		}
+
+		err := getTLS(ctx, ingress, client, tlsConfigs)
+		if err != nil {
+			log.FromContext(ctx).Errorf("Error configuring TLS: %v", err)
+		}
+
+		if len(ingress.Spec.Rules) == 0 {
```

Comment:
You can have `ingress.Spec.Rules` _and_ `ingress.Spec.Backend` at the same time. Not sure this if is correct.